src/Thorn_Calculus-Core_Bisimilarities.thy

```diff
@@ -1469,9 +1467,177 @@ proof -
   finally show ?thesis .
 qed
 
+private lemma adapted_after_repeated_receive_nested_idempotency:
+  shows "(A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> \<E> \<parallel> ((A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> \<E> \<parallel> P) \<parallel> Q \<guillemotleft> \<E> \<sim>\<^sub>s P \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x \<parallel> Q) \<guillemotleft> \<E>"
```

Comment:
Like the statements of `double_repeated_receive_post_left_receive` and `double_repeated_receive_post_right_receive`, this lemma statement and the next should be more specific, I think. I’d prefer if they stated exactly what’s needed in the proof of `inner_repeated_receive_redundancy`. Not only would this better motivate the introduction of these lemmas; it would also remove sources of confusion. Currently, there are unfortunate mismatches between the variable names in the lemma statements and the ones in the proof of `inner_repeated_receive_redundancy`; for example, `P` in the statement above gets instantiated with `post_receive n X 𝒬`.

Of course, the lemma statements get longer this way. I strongly suggest to use `is` annotations to avoid repeating long terms in the proofs.

Reply:
Fixed in 9e401d2987a9a677cbb2ae49f569bf1d1f7f7a88, please verify.

Reply:
I forgot to include the removal of superfluous parentheses in the previous commit, so I created ff75591965a0daa1f277b886be9aac45c625528a to fix it.

Reply:
https://github.com/input-output-hk/thorn-calculus/blob/9e401d2987a9a677cbb2ae49f569bf1d1f7f7a88/src/Thorn_Calculus-Core_Bisimilarities.thy#L1476
`T` and `U` should be turned into `S` and `T`. `S` and `T` are often used for a left-hand and corresponding right-hand side. `U` is typically used on its own.

Reply:
https://github.com/input-output-hk/thorn-calculus/blob/9e401d2987a9a677cbb2ae49f569bf1d1f7f7a88/src/Thorn_Calculus-Core_Bisimilarities.thy#L1478-L1480
This should become a single line, analogously to what we already have in many places in [network-equivalences](https://github.com/input-output-hk/network-equivalences) with `…` instead of `?T`.

Reply:
In other places, we write parallel composition chains without parentheses,  like `P ∥ Q ∥ R`, only where we don’t have any particular nesting in mind. The idea is that the terms we write would stay valid if we changed the associativity of `∥`.

However, the left-hand side of the first of the auxiliary lemmas and the right-hand side of the second must have the nesting structure `P ∥ (Q ∥ R)`, since each arises from having a parallel composition perform a transition, with `P` arising from the left subprocess of that parallel composition and `Q ∥ R` arising from the right.

Therefore, I ask you to add parentheses around the respective processes `Q ∥ R`. In the case of the first one, please move the line break that splits the whole left-hand side such that it ends up just before the parallel composition `Q ∥ R`.

Here are the source code snippets that contain the two processes that should get parentheses added:
https://github.com/input-output-hk/thorn-calculus/blob/9e401d2987a9a677cbb2ae49f569bf1d1f7f7a88/src/Thorn_Calculus-Core_Bisimilarities.thy#L1472-L1473
https://github.com/input-output-hk/thorn-calculus/blob/9e401d2987a9a677cbb2ae49f569bf1d1f7f7a88/src/Thorn_Calculus-Core_Bisimilarities.thy#L1495

Reply:
> https://github.com/input-output-hk/thorn-calculus/blob/9e401d2987a9a677cbb2ae49f569bf1d1f7f7a88/src/Thorn_Calculus-Core_Bisimilarities.thy#L1476
> 
> `T` and `U` should be turned into `S` and `T`. `S` and `T` are often used for a left-hand and corresponding right-hand side. `U` is typically used on its own.

Fixed in 16c0e8e0b1f29d144bc25fa098c9504d8fc79917, please verify.

Reply:
> https://github.com/input-output-hk/thorn-calculus/blob/9e401d2987a9a677cbb2ae49f569bf1d1f7f7a88/src/Thorn_Calculus-Core_Bisimilarities.thy#L1478-L1480
> 
> This should become a single line, analogously to what we already have in many places in [network-equivalences](https://github.com/input-output-hk/network-equivalences) with `…` instead of `?T`.

Fixed in 16c0e8e0b1f29d144bc25fa098c9504d8fc79917, please verify.

Reply:
> In other places, we write parallel composition chains without parentheses, like `P ∥ Q ∥ R`, only where we don’t have any particular nesting in mind. The idea is that the terms we write would stay valid if we changed the associativity of `∥`.
> 
> However, the left-hand side of the first of the auxiliary lemmas and the right-hand side of the second must have the nesting structure `P ∥ (Q ∥ R)`, since each arises from having a parallel composition perform a transition, with `P` arising from the left subprocess of that parallel composition and `Q ∥ R` arising from the right.
> 
> Therefore, I ask you to add parentheses around the respective processes `Q ∥ R`. In the case of the first one, please move the line break that splits the whole left-hand side such that it ends up just before the parallel composition `Q ∥ R`.
> 
> Here are the source code snippets that contain the two processes that should get parentheses added:
> 
> https://github.com/input-output-hk/thorn-calculus/blob/9e401d2987a9a677cbb2ae49f569bf1d1f7f7a88/src/Thorn_Calculus-Core_Bisimilarities.thy#L1472-L1473
> 
> https://github.com/input-output-hk/thorn-calculus/blob/9e401d2987a9a677cbb2ae49f569bf1d1f7f7a88/src/Thorn_Calculus-Core_Bisimilarities.thy#L1495

Fixed in 33181bbf6697437e552b24117dfee6c338c84878, please verify.